ice: auto traits: assertion failed `1 == 2`

<!--
ICE: Rustc ./84D93BBD96EACE9F64B0DDBE84C1EBC57C106D78FAE04E5F695D0504C2F603B4.rs '' 'thread 'rustc' panicked at compiler/rustc_hir_analysis/src/coherence/orphan.rs:455:5: 'assertion `left == right` failed'', 'thread 'rustc' panicked at compiler/rustc_hir_analysis/src/coherence/orphan.rs:455:5: 'assertion `left == right` failed''
File: /tmp/icemaker/84D93BBD96EACE9F64B0DDBE84C1EBC57C106D78FAE04E5F695D0504C2F603B4.rs
-->

auto-reduced (treereduce-rust):
````rust




auto trait Trait<P> {


}

impl<P> Trait<P> for () {
     ;
}



auto trait Bound : Copy {}

auto trait LifetimeBound : 'static {}

auto trait MyTrait {  }




````

<details><summary><strong>original code</strong></summary>
<p>

original:
````rust
#![feature(auto_traits)]

// run-rustfix

auto trait Trait<P> {
    const A: ();
    const B<const K: u64, const Q: u64>: u64;
    const C<T>: T;
    const D<const N: usize>: usize;

    const E: usize;
    const F<T: PartialEq>: ();
}

impl<P> Trait<P> for () {
    const A<T>: () = ();
    //~^ ERROR const `A` has 1 type parameter but its trait declaration has 0 type parameters
    const B<const K: u64>: u64 = 0;
    //~^ ERROR const `B` has 1 const parameter but its trait declaration has 2 const parameters
    const C<'a>: &'a str = "";
    //~^ ERROR const `C` has 0 type parameters but its trait declaration has 1 type parameter
    const D<const N: u16>: u16 = N;
    //~^ ERROR const `D` has an incompatible generic parameter for trait `Trait`

    const E: usize = 0x11111111
    where
        P: Copy; //~ ERROR impl has stricter requirements than trait
    const F<T: Eq>: () = (); //~ ERROR impl has stricter requirements than trait
}

fn main() {}
//~^ auto traits cannot have generic parameters [E0567]
auto trait Bound : Copy {}
//~^ auto traits cannot have super traits or lifetime bounds [E0568]
auto trait LifetimeBound : 'static {}
//~^ auto traits cannot have super traits or lifetime bounds [E0568]
auto trait MyTrait { fn foo() {} }
//~^ auto traits cannot have associated items [E0380]
fn main() {}

````
</p>
</details>

Version information
````
rustc 1.75.0-nightly (17d0a45f5 2023-11-10)
binary: rustc
commit-hash: 17d0a45f5df31f218b84c3275bb05c3227348b5b
commit-date: 2023-11-10
host: x86_64-unknown-linux-gnu
release: 1.75.0-nightly
LLVM version: 17.0.4

````

Command:
`/home/matthias/.rustup/toolchains/master/bin/rustc `

<!--
Include a backtrace in the code block by setting `RUST_BACKTRACE=1` in your
environment. E.g. `RUST_BACKTRACE=1 cargo build`.
-->
<details><summary><strong>Program output</strong></summary>
<p>

```
error: non-item in item list
  --> /tmp/icemaker_global_tempdir.lcHIbosRSaxE/rustc_testrunner_tmpdir_reporting.aKyC4RFSO5Er/mvce.rs:11:6
   |
10 | impl<P> Trait<P> for () {
   |                         - item list starts here
11 |      ;
   |      ^ non-item starts here
12 | }
   | - item list ends here

error[E0567]: auto traits cannot have generic parameters
 --> /tmp/icemaker_global_tempdir.lcHIbosRSaxE/rustc_testrunner_tmpdir_reporting.aKyC4RFSO5Er/mvce.rs:5:17
  |
5 | auto trait Trait<P> {
  |            -----^^^ help: remove the parameters
  |            |
  |            auto trait cannot have generic parameters

error[E0568]: auto traits cannot have super traits or lifetime bounds
  --> /tmp/icemaker_global_tempdir.lcHIbosRSaxE/rustc_testrunner_tmpdir_reporting.aKyC4RFSO5Er/mvce.rs:16:17
   |
16 | auto trait Bound : Copy {}
   |            -----^^^^^^^ help: remove the super traits or lifetime bounds
   |            |
   |            auto traits cannot have super traits or lifetime bounds

error[E0568]: auto traits cannot have super traits or lifetime bounds
  --> /tmp/icemaker_global_tempdir.lcHIbosRSaxE/rustc_testrunner_tmpdir_reporting.aKyC4RFSO5Er/mvce.rs:18:25
   |
18 | auto trait LifetimeBound : 'static {}
   |            -------------^^^^^^^^^^ help: remove the super traits or lifetime bounds
   |            |
   |            auto traits cannot have super traits or lifetime bounds

error[E0658]: auto traits are experimental and possibly buggy
 --> /tmp/icemaker_global_tempdir.lcHIbosRSaxE/rustc_testrunner_tmpdir_reporting.aKyC4RFSO5Er/mvce.rs:5:1
  |
5 | / auto trait Trait<P> {
6 | |
7 | |
8 | | }
  | |_^
  |
  = note: see issue #13231 <https://github.com/rust-lang/rust/issues/13231> for more information
  = help: add `#![feature(auto_traits)]` to the crate attributes to enable

error[E0658]: auto traits are experimental and possibly buggy
  --> /tmp/icemaker_global_tempdir.lcHIbosRSaxE/rustc_testrunner_tmpdir_reporting.aKyC4RFSO5Er/mvce.rs:16:1
   |
16 | auto trait Bound : Copy {}
   | ^^^^^^^^^^^^^^^^^^^^^^^^^^
   |
   = note: see issue #13231 <https://github.com/rust-lang/rust/issues/13231> for more information
   = help: add `#![feature(auto_traits)]` to the crate attributes to enable

error[E0658]: auto traits are experimental and possibly buggy
  --> /tmp/icemaker_global_tempdir.lcHIbosRSaxE/rustc_testrunner_tmpdir_reporting.aKyC4RFSO5Er/mvce.rs:18:1
   |
18 | auto trait LifetimeBound : 'static {}
   | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
   |
   = note: see issue #13231 <https://github.com/rust-lang/rust/issues/13231> for more information
   = help: add `#![feature(auto_traits)]` to the crate attributes to enable

error[E0658]: auto traits are experimental and possibly buggy
  --> /tmp/icemaker_global_tempdir.lcHIbosRSaxE/rustc_testrunner_tmpdir_reporting.aKyC4RFSO5Er/mvce.rs:20:1
   |
20 | auto trait MyTrait {  }
   | ^^^^^^^^^^^^^^^^^^^^^^^
   |
   = note: see issue #13231 <https://github.com/rust-lang/rust/issues/13231> for more information
   = help: add `#![feature(auto_traits)]` to the crate attributes to enable

error[E0601]: `main` function not found in crate `mvce`
  --> /tmp/icemaker_global_tempdir.lcHIbosRSaxE/rustc_testrunner_tmpdir_reporting.aKyC4RFSO5Er/mvce.rs:20:24
   |
20 | auto trait MyTrait {  }
   |                        ^ consider adding a `main` function to `/tmp/icemaker_global_tempdir.lcHIbosRSaxE/rustc_testrunner_tmpdir_reporting.aKyC4RFSO5Er/mvce.rs`

thread 'rustc' panicked at compiler/rustc_hir_analysis/src/coherence/orphan.rs:455:5:
assertion `left == right` failed
  left: 2
 right: 1
stack backtrace:
   0:     0x7fb17d38b62c - std::backtrace_rs::backtrace::libunwind::trace::h46374b84341786bc
                               at /rustc/17d0a45f5df31f218b84c3275bb05c3227348b5b/library/std/src/../../backtrace/src/backtrace/libunwind.rs:104:5
   1:     0x7fb17d38b62c - std::backtrace_rs::backtrace::trace_unsynchronized::he7815f30c698ee16
                               at /rustc/17d0a45f5df31f218b84c3275bb05c3227348b5b/library/std/src/../../backtrace/src/backtrace/mod.rs:66:5
   2:     0x7fb17d38b62c - std::sys_common::backtrace::_print_fmt::hf3ac027a6e20df81
                               at /rustc/17d0a45f5df31f218b84c3275bb05c3227348b5b/library/std/src/sys_common/backtrace.rs:67:5
   3:     0x7fb17d38b62c - <std::sys_common::backtrace::_print::DisplayBacktrace as core::fmt::Display>::fmt::h6350b6493143b1fc
                               at /rustc/17d0a45f5df31f218b84c3275bb05c3227348b5b/library/std/src/sys_common/backtrace.rs:44:22
   4:     0x7fb17d3de630 - core::fmt::rt::Argument::fmt::haac38874886df8e7
                               at /rustc/17d0a45f5df31f218b84c3275bb05c3227348b5b/library/core/src/fmt/rt.rs:142:9
   5:     0x7fb17d3de630 - core::fmt::write::hf0e822b6d8ad56b2
                               at /rustc/17d0a45f5df31f218b84c3275bb05c3227348b5b/library/core/src/fmt/mod.rs:1120:17
   6:     0x7fb17d37f54f - std::io::Write::write_fmt::h1a44df357aa13b6e
                               at /rustc/17d0a45f5df31f218b84c3275bb05c3227348b5b/library/std/src/io/mod.rs:1762:15
   7:     0x7fb17d38b414 - std::sys_common::backtrace::_print::h95329df2b370286a
                               at /rustc/17d0a45f5df31f218b84c3275bb05c3227348b5b/library/std/src/sys_common/backtrace.rs:47:5
   8:     0x7fb17d38b414 - std::sys_common::backtrace::print::hd1549e5974c3fa44
                               at /rustc/17d0a45f5df31f218b84c3275bb05c3227348b5b/library/std/src/sys_common/backtrace.rs:34:9
   9:     0x7fb17d38e0a7 - std::panicking::default_hook::{{closure}}::hfc23d70499246b08
  10:     0x7fb17d38de0f - std::panicking::default_hook::h0d66df0a095a3bd0
                               at /rustc/17d0a45f5df31f218b84c3275bb05c3227348b5b/library/std/src/panicking.rs:292:9
  11:     0x7fb1800ee150 - std[4df474cf7c0204ae]::panicking::update_hook::<alloc[ca626be192fcf572]::boxed::Box<rustc_driver_impl[ec4025a53c2103e9]::install_ice_hook::{closure#0}>>::{closure#0}
  12:     0x7fb17d38e7e8 - <alloc::boxed::Box<F,A> as core::ops::function::Fn<Args>>::call::h44f7f870300bcc1d
                               at /rustc/17d0a45f5df31f218b84c3275bb05c3227348b5b/library/alloc/src/boxed.rs:2021:9
  13:     0x7fb17d38e7e8 - std::panicking::rust_panic_with_hook::h652875067b2fe609
                               at /rustc/17d0a45f5df31f218b84c3275bb05c3227348b5b/library/std/src/panicking.rs:783:13
  14:     0x7fb17d38e53e - std::panicking::begin_panic_handler::{{closure}}::h441bd00acf2d3372
                               at /rustc/17d0a45f5df31f218b84c3275bb05c3227348b5b/library/std/src/panicking.rs:657:13
  15:     0x7fb17d38baf6 - std::sys_common::backtrace::__rust_end_short_backtrace::h302c26748ffe0429
                               at /rustc/17d0a45f5df31f218b84c3275bb05c3227348b5b/library/std/src/sys_common/backtrace.rs:170:18
  16:     0x7fb17d38e2a2 - rust_begin_unwind
                               at /rustc/17d0a45f5df31f218b84c3275bb05c3227348b5b/library/std/src/panicking.rs:645:5
  17:     0x7fb17d3dad15 - core::panicking::panic_fmt::heeeb5ab19bc3d1b2
                               at /rustc/17d0a45f5df31f218b84c3275bb05c3227348b5b/library/core/src/panicking.rs:72:14
  18:     0x7fb17d3db28b - core::panicking::assert_failed_inner::hc77e496884c61372
  19:     0x7fb1801bcf03 - core[f248d18e5721e18e]::panicking::assert_failed::<usize, usize>
  20:     0x7fb18167fb07 - rustc_hir_analysis[3fdc3bc95e806e7]::coherence::orphan::orphan_check_impl
  21:     0x7fb18167e719 - rustc_query_impl[d9d242dab8718f47]::plumbing::__rust_begin_short_backtrace::<rustc_query_impl[d9d242dab8718f47]::query_impl::orphan_check_impl::dynamic_query::{closure#2}::{closure#0}, rustc_middle[dadee63cb0feb6d5]::query::erase::Erased<[u8; 1usize]>>
  22:     0x7fb181681bba - rustc_query_system[6b6f1c30493de844]::query::plumbing::try_execute_query::<rustc_query_impl[d9d242dab8718f47]::DynamicConfig<rustc_query_system[6b6f1c30493de844]::query::caches::VecCache<rustc_span[9afbaaf51cc05e3]::def_id::LocalDefId, rustc_middle[dadee63cb0feb6d5]::query::erase::Erased<[u8; 1usize]>>, false, false, false>, rustc_query_impl[d9d242dab8718f47]::plumbing::QueryCtxt, false>
  23:     0x7fb1816817c3 - rustc_query_impl[d9d242dab8718f47]::query_impl::orphan_check_impl::get_query_non_incr::__rust_end_short_backtrace
  24:     0x7fb1818b032a - rustc_hir_analysis[3fdc3bc95e806e7]::coherence::coherent_trait
  25:     0x7fb1818afeb9 - rustc_query_impl[d9d242dab8718f47]::plumbing::__rust_begin_short_backtrace::<rustc_query_impl[d9d242dab8718f47]::query_impl::coherent_trait::dynamic_query::{closure#2}::{closure#0}, rustc_middle[dadee63cb0feb6d5]::query::erase::Erased<[u8; 0usize]>>
  26:     0x7fb181e0dfbd - rustc_query_system[6b6f1c30493de844]::query::plumbing::try_execute_query::<rustc_query_impl[d9d242dab8718f47]::DynamicConfig<rustc_query_system[6b6f1c30493de844]::query::caches::DefaultCache<rustc_span[9afbaaf51cc05e3]::def_id::DefId, rustc_middle[dadee63cb0feb6d5]::query::erase::Erased<[u8; 0usize]>>, false, false, false>, rustc_query_impl[d9d242dab8718f47]::plumbing::QueryCtxt, false>
  27:     0x7fb181e0dd4d - rustc_query_impl[d9d242dab8718f47]::query_impl::coherent_trait::get_query_non_incr::__rust_end_short_backtrace
  28:     0x7fb1819ce469 - rustc_hir_analysis[3fdc3bc95e806e7]::check_crate
  29:     0x7fb1817aa8d7 - rustc_interface[b3bf249fff60da9a]::passes::analysis
  30:     0x7fb1817aa537 - rustc_query_impl[d9d242dab8718f47]::plumbing::__rust_begin_short_backtrace::<rustc_query_impl[d9d242dab8718f47]::query_impl::analysis::dynamic_query::{closure#2}::{closure#0}, rustc_middle[dadee63cb0feb6d5]::query::erase::Erased<[u8; 1usize]>>
  31:     0x7fb182398824 - rustc_query_system[6b6f1c30493de844]::query::plumbing::try_execute_query::<rustc_query_impl[d9d242dab8718f47]::DynamicConfig<rustc_query_system[6b6f1c30493de844]::query::caches::SingleCache<rustc_middle[dadee63cb0feb6d5]::query::erase::Erased<[u8; 1usize]>>, false, false, false>, rustc_query_impl[d9d242dab8718f47]::plumbing::QueryCtxt, false>
  32:     0x7fb182398615 - rustc_query_impl[d9d242dab8718f47]::query_impl::analysis::get_query_non_incr::__rust_end_short_backtrace
  33:     0x7fb1823a123f - rustc_interface[b3bf249fff60da9a]::interface::run_compiler::<core[f248d18e5721e18e]::result::Result<(), rustc_span[9afbaaf51cc05e3]::ErrorGuaranteed>, rustc_driver_impl[ec4025a53c2103e9]::run_compiler::{closure#1}>::{closure#0}
  34:     0x7fb1823a855a - std[4df474cf7c0204ae]::sys_common::backtrace::__rust_begin_short_backtrace::<rustc_interface[b3bf249fff60da9a]::util::run_in_thread_with_globals<rustc_interface[b3bf249fff60da9a]::util::run_in_thread_pool_with_globals<rustc_interface[b3bf249fff60da9a]::interface::run_compiler<core[f248d18e5721e18e]::result::Result<(), rustc_span[9afbaaf51cc05e3]::ErrorGuaranteed>, rustc_driver_impl[ec4025a53c2103e9]::run_compiler::{closure#1}>::{closure#0}, core[f248d18e5721e18e]::result::Result<(), rustc_span[9afbaaf51cc05e3]::ErrorGuaranteed>>::{closure#0}, core[f248d18e5721e18e]::result::Result<(), rustc_span[9afbaaf51cc05e3]::ErrorGuaranteed>>::{closure#0}::{closure#0}, core[f248d18e5721e18e]::result::Result<(), rustc_span[9afbaaf51cc05e3]::ErrorGuaranteed>>
  35:     0x7fb1823a8383 - <<std[4df474cf7c0204ae]::thread::Builder>::spawn_unchecked_<rustc_interface[b3bf249fff60da9a]::util::run_in_thread_with_globals<rustc_interface[b3bf249fff60da9a]::util::run_in_thread_pool_with_globals<rustc_interface[b3bf249fff60da9a]::interface::run_compiler<core[f248d18e5721e18e]::result::Result<(), rustc_span[9afbaaf51cc05e3]::ErrorGuaranteed>, rustc_driver_impl[ec4025a53c2103e9]::run_compiler::{closure#1}>::{closure#0}, core[f248d18e5721e18e]::result::Result<(), rustc_span[9afbaaf51cc05e3]::ErrorGuaranteed>>::{closure#0}, core[f248d18e5721e18e]::result::Result<(), rustc_span[9afbaaf51cc05e3]::ErrorGuaranteed>>::{closure#0}::{closure#0}, core[f248d18e5721e18e]::result::Result<(), rustc_span[9afbaaf51cc05e3]::ErrorGuaranteed>>::{closure#1} as core[f248d18e5721e18e]::ops::function::FnOnce<()>>::call_once::{shim:vtable#0}
  36:     0x7fb17d3986a5 - <alloc::boxed::Box<F,A> as core::ops::function::FnOnce<Args>>::call_once::h473dfddbf197f5bf
                               at /rustc/17d0a45f5df31f218b84c3275bb05c3227348b5b/library/alloc/src/boxed.rs:2007:9
  37:     0x7fb17d3986a5 - <alloc::boxed::Box<F,A> as core::ops::function::FnOnce<Args>>::call_once::h8959fbdf8cbf8098
                               at /rustc/17d0a45f5df31f218b84c3275bb05c3227348b5b/library/alloc/src/boxed.rs:2007:9
  38:     0x7fb17d3986a5 - std::sys::unix::thread::Thread::new::thread_start::hab410919d7ef6e45
                               at /rustc/17d0a45f5df31f218b84c3275bb05c3227348b5b/library/std/src/sys/unix/thread.rs:108:17
  39:     0x7fb17d1899eb - <unknown>
  40:     0x7fb17d20d7cc - <unknown>
  41:                0x0 - <unknown>

error: the compiler unexpectedly panicked. this is a bug.

note: we would appreciate a bug report: https://github.com/rust-lang/rust/issues/new?labels=C-bug%2C+I-ICE%2C+T-compiler&template=ice.md

note: rustc 1.75.0-nightly (17d0a45f5 2023-11-10) running on x86_64-unknown-linux-gnu

query stack during panic:
#0 [orphan_check_impl] checking whether impl `<impl at /tmp/icemaker_global_tempdir.lcHIbosRSaxE/rustc_testrunner_tmpdir_reporting.aKyC4RFSO5Er/mvce.rs:10:1: 10:24>` follows the orphan rules
#1 [coherent_trait] coherence checking all impls of trait `Trait`
#2 [analysis] running analysis passes on this crate
end of query stack
error: aborting due to 9 previous errors

Some errors have detailed explanations: E0567, E0568, E0601, E0658.
For more information about an error, try `rustc --explain E0567`.

```

</p>
</details>

--- a/compiler/rustc_hir_analysis/src/coherence/orphan.rs
+++ b/compiler/rustc_hir_analysis/src/coherence/orphan.rs
@@ -452,7 +452,13 @@ fn lint_auto_trait_impl<'tcx>(
     trait_ref: ty::TraitRef<'tcx>,
     impl_def_id: LocalDefId,
 ) {
-    assert_eq!(trait_ref.args.len(), 1);
+    if trait_ref.args.len() != 1 {
+        tcx.sess.diagnostic().delay_span_bug(
+            tcx.def_span(impl_def_id),
+            "auto traits cannot have generic parameters",
+        );
+        return;
+    }
     let self_ty = trait_ref.self_ty();
     let (self_type_did, args) = match self_ty.kind() {
         ty::Adt(def, args) => (def.did(), args),

--- a/tests/ui/auto-traits/issue-117789.rs
+++ b/tests/ui/auto-traits/issue-117789.rs
@@ -1,0 +1,7 @@
+#![deny(suspicious_auto_trait_impls)]
+
+auto trait Trait<P> {} //~ ERROR auto traits cannot have generic parameters
+//~^ ERROR auto traits are experimental and possibly buggy
+impl<P> Trait<P> for () {}
+
+fn main() {}

--- a/tests/ui/auto-traits/issue-117789.stderr
+++ b/tests/ui/auto-traits/issue-117789.stderr
@@ -1,0 +1,21 @@
+error[E0567]: auto traits cannot have generic parameters
+  --> $DIR/issue-117789.rs:3:17
+   |
+LL | auto trait Trait<P> {}
+   |            -----^^^ help: remove the parameters
+   |            |
+   |            auto trait cannot have generic parameters
+
+error[E0658]: auto traits are experimental and possibly buggy
+  --> $DIR/issue-117789.rs:3:1
+   |
+LL | auto trait Trait<P> {}
+   | ^^^^^^^^^^^^^^^^^^^^^^
+   |
+   = note: see issue #13231 <https://github.com/rust-lang/rust/issues/13231> for more information
+   = help: add `#![feature(auto_traits)]` to the crate attributes to enable
+
+error: aborting due to 2 previous errors
+
+Some errors have detailed explanations: E0567, E0658.
+For more information about an error, try `rustc --explain E0567`.

